Sources/MapboxNavigation/NavigationViewportDataSource.swift

```diff
@@ -213,18 +213,31 @@ public class NavigationViewportDataSource: ViewportDataSource {
             if geometryFramingAfterManeuver.enabled {
                 let stepIndex = routeProgress.currentLegProgress.stepIndex
                 let nextStepIndex = min(stepIndex + 1, routeProgress.currentLeg.steps.count - 1)
-                let stepCoordinatesAfterCurrentStep = routeProgress.currentLeg.steps[nextStepIndex...]
-                    .map({ $0.shape?.coordinates })
                 
-                for stepCoordinates in stepCoordinatesAfterCurrentStep {
-                    guard let stepCoordinates = stepCoordinates,
+                var totalDistance: CLLocationDistance = 0.0
+                for (index, step) in routeProgress.currentLeg.steps.suffix(from: nextStepIndex).enumerated() {
+                    guard let stepCoordinates = step.shape?.coordinates,
                           let distance = stepCoordinates.distance() else { continue }
                     
-                    if distance > 0.0 && distance < geometryFramingAfterManeuver.distanceToCoalesceCompoundManeuvers {
-                        compoundManeuvers.append(stepCoordinates)
-                    } else {
-                        compoundManeuvers.append(stepCoordinates.trimmed(distance: geometryFramingAfterManeuver.distanceToFrameAfterManeuver))
-                        break
+                    if index == 0 {
+                        if distance >= geometryFramingAfterManeuver.distanceToFrameAfterManeuver {
+                            let trimmedStepCoordinates = stepCoordinates.trimmed(distance: geometryFramingAfterManeuver.distanceToFrameAfterManeuver)
+                            compoundManeuvers.append(trimmedStepCoordinates)
+                            break
+                        } else {
+                            compoundManeuvers.append(stepCoordinates)
+                            totalDistance += distance
+                        }
+                    } else if distance >= 0.0 && totalDistance < geometryFramingAfterManeuver.distanceToCoalesceCompoundManeuvers {
+                        if distance + totalDistance >= geometryFramingAfterManeuver.distanceToCoalesceCompoundManeuvers {
+                            let remanentDistance = geometryFramingAfterManeuver.distanceToCoalesceCompoundManeuvers - totalDistance
+                            let trimmedStepCoordinates = stepCoordinates.trimmed(distance: remanentDistance)
+                            compoundManeuvers.append(trimmedStepCoordinates)
+                            break
+                        } else {
+                            compoundManeuvers.append(stepCoordinates)
+                            totalDistance += distance
```

Comment:
By the way, similar logic in `Route.polylineAroundManeuver(legIndex:stepIndex:distance:)` would probably mitigate https://github.com/mapbox/mapbox-navigation-ios/issues/3589#issuecomment-968192122 and reduce the need for #3638.

Reply:
I think the pencil tip maneuver should be handled separately.  Because in the complex intersection issue, we limit the overlooking distance for camera based on different step and step coordinates. But in the pencil tip maneuver issue, the cause is that the maneuver is pointed at the first turn, while the second turn is included in the next step. There's no step difference for the next turn. So we should have the overlooking distance for maneuver by the `turn` instead of by the step index or step distance. The identification of `turn` from shape should be addressed in this situation instead of `step`